### PR TITLE
Merge module for Convolution

### PR DESCRIPTION
**Pull request by stelladk**

### CODE REVIEW

File: src/gromo/modules/conv2d_growing_module.py [hunk 1]
```diff
@@ -69,7 +218,6 @@ def __init__(
         name: str | None = None,
         s_growth_is_needed: bool = False,
     ) -> None:
-        device = device if device is not None else global_device()
```
Comment (sylvchev) [inline src/gromo/modules/conv2d_growing_module.py:72]: Why removing this line?

**Review by sylvchev — COMMENTED** [review]
Nice addition, it seems ok for merge module. I'm not sure about the drop global device, why removing it?

File: src/gromo/modules/conv2d_growing_module.py [hunk 2]
```diff
@@ -17,7 +17,156 @@
...
+            )
+        else:
+            self.previous_tensor_m = None
+
+    def construct_full_activity(self) -> torch.Tensor:
+        assert self.previous_modules, f"No previous modules for {self.name}."
+        full_activity = torch.ones(
+            (self.previous_modules[0].input.shape[0], self.total_in_features),
+            device=self.device,
+        )
```
Comment (TheoRudkiewicz) [inline src/gromo/modules/conv2d_growing_module.py:145]: I think this wont work. You create a 2-order tensor while expecting a 3-order tensor when computing S. I think `full_activity` should be of shape `(n, total_in_features, out_channels)`. (where n is the number of examples in the batch).
Anyway, if I am not mistaken you will immediately see it with the sanity checks.

File: src/gromo/modules/conv2d_growing_module.py [hunk 3]
```diff
@@ -120,6 +268,35 @@ def padding(self):
...
+        return self.layer.stride
+
+    def __out_dimention(self, dim: int) -> int:
+        return (
+            int(
+                (self.input_size[dim] - self.kernel_size[dim] + 2 * self.padding[dim])
+                / self.stride[dim]
+            )
+            + 1
+        )
```
Comment (TheoRudkiewicz) [inline src/gromo/modules/conv2d_growing_module.py:282]: If you want you could use the function `compute_output_shape_conv` from `gromo.utils.tools` that does already (almost) those computation and has been tested.

File: src/gromo/modules/conv2d_growing_module.py [hunk 4]
```diff
@@ -120,6 +268,35 @@ def padding(self):
...
+    def out_width(self) -> int:
+        return self.__out_dimention(0)
+
+    @property
+    def out_height(self) -> int:
+        return self.__out_dimention(1)
+
+    @property
+    def in_features(self) -> int:
+        return self.in_channels * self.input_size[0] * self.input_size[1] + self.use_bias
```
Comment (TheoRudkiewicz) [inline src/gromo/modules/conv2d_growing_module.py:294]: Very nice and quiet useful !

**Review by TheoRudkiewicz — COMMENTED** [review]
It's looking good. If I may I would suggest to do a first and separate pull request where we switch `compute_optimal_delta` form child class to `MergeModule` thus this pull request would not change `growing_module` nor `linear_growing_module`. (I could do this separate PR if you want)

File: src/gromo/modules/conv2d_growing_module.py (hunk 2 above)
Comment (stelladk) [reply src/gromo/modules/conv2d_growing_module.py:146]: Yes, you are right! But (n, total_in_features, out_channels) also doesn't match. The unfolded_extended_input used below is of shape (n, module.in_channels*P{module.kernel_size}, L} if I'm not mistaken. https://docs.pytorch.org/docs/stable/generated/torch.nn.Unfold.html#torch.nn.Unfold
Is this really what we want, or should we split it into modules in some other way?

File: src/gromo/modules/conv2d_growing_module.py [hunk 5]
```diff
@@ -18,7 +18,199 @@
...
+        flattened = self.activity.view(batch_size, -1)  # (N, C⋅H⋅W)
+
+        if self.use_bias:
+            ones = torch.ones(batch_size, 1, device=self.activity.device)
+            input_extended = torch.cat((flattened, ones), dim=1)  # (N, C⋅H⋅W + 1)
+            update = torch.einsum("ij,ik->jk", input_extended, input_extended)
+        else:
+            update = torch.einsum("ij,ik->jk", flattened, flattened)
+
+        return update, batch_size
```
Comment (stelladk) [inline src/gromo/modules/conv2d_growing_module.py:213]: @TheoRudkiewicz Could you check if this method is correct?

Comment (TheoRudkiewicz) [reply src/gromo/modules/conv2d_growing_module.py:213]: I don't think it works because the update should be of shape `(total_in_features, total_in_features)` which is $C \times H_d \times W_d$ where $(H_d, W_d)$ is the kernel size. In this code the update will be $C \times H \times W$  where $(H, W)$ is the "spatial" shape of the image.

Comment (stelladk) [reply src/gromo/modules/conv2d_growing_module.py]: After in-person discussion I have corrected the shape of tensor_s in Conv2dMergeGrowingModule and handled compute_s_update with the unfold function

**Comment by stelladk** [comment]
> @stelladk Is the `compute_s_update` good now?

Yes, it is working correctly! However, there is still more testing to be done on this PR.

File: src/gromo/modules/conv2d_growing_module.py [hunk 6]
```diff
@@ -14,11 +14,382 @@
...
+            if module.out_channels != self.in_channels:
+                raise ValueError(
+                    "The input channels must match the output channels of the previous modules."
+                )
+            self.total_in_features += module.in_features + module.use_bias
+            self.total_out_features += module.out_features
+            # self.total_out_features += module.use_bias
+        if self.total_in_features > 0:
+            if self.input_size is None:
+                self.input_size = (module.out_width, module.out_height)
```
Comment (stephane-rivaud) [inline src/gromo/modules/conv2d_growing_module.py:224]: We should not use `module` outside of the for loop. This is not supposed to happen.
```suggestion
                self.input_size = (self.previous_module[0].out_width, self.previous_module[0].out_height)
```

**Comment by stephane-rivaud** [comment]
# PR #94: Implement Conv2dMergeGrowingModule - Comprehensive Analysis

## 🎯 **Objective**
This PR implements the `Conv2dMergeGrowingModule` class, a critical component for merging convolutional neural network layers in the GROMO (Growing Neural Networks) framework. This implementation enables dynamic network growth by allowing multiple convolutional layers to be merged into a single computational unit.

## 📊 **Summary Statistics**
- **Total changes**: 1,769 insertions, 146 deletions across 10 files
- **Core implementation**: 452+ lines of new functionality in `Conv2dMergeGrowingModule`
- **Test coverage**: 595+ lines of comprehensive tests
- **Coverage achievement**: 100% diff coverage for newly introduced code
- **Files modified**: 10 files (7 source files, 3 test files, 1 documentation file)

---

## 🔧 **Core Implementation: Conv2dMergeGrowingModule**

### **Key Features Implemented**

#### **1. Constructor & Initialization**
```python
def __init__(
    self,
    in_channels: int,
    input_size: int | tuple[int, int],
    next_kernel_size: int | tuple[int, int],
    post_merge_function: torch.nn.Module = torch.nn.Identity(),
    previous_modules: list[GrowingModule | MergeGrowingModule] | None = None,
    next_modules: list[GrowingModule | MergeGrowingModule] | None = None,
    allow_growing: bool = False,
    input_volume: int | None = None,
    device: torch.device | None = None,
    name: str | None = None,
) -> None:
```

**Features:**
- ✅ Flexible parameter types (int or tuple for sizes)
- ✅ Automatic conversion of single integers to tuples
- ✅ Integration with parent `MergeGrowingModule` class
- ✅ Proper tensor shape initialization for statistical tracking

#### **2. Properties & Metadata**
- **`input_volume`**: Dynamic calculation based on previous modules or explicit override
- **`out_channels`**, **`in_features`**, **`out_features`**: Channel and feature dimension management
- **`output_size`**: Spatial dimension preservation
- **`padding`**, **`stride`**, **`dilation`**: Dynamic property derivation from next modules

#### **3. Module Connection Management**

**`set_next_modules()`**:
- ✅ Validates that next modules are `Conv2dGrowingModule` or `LinearGrowingModule`
- ✅ Enforces type consistency across all next modules
- ✅ Validates kernel size compatibility for conv modules
- ✅ Automatically updates connected module properties

**`set_previous_modules()`**:
- ✅ Validates that previous modules are `Conv2dGrowingModule` only
- ✅ Enforces kernel size consistency
- ✅ Validates channel compatibility
- ✅ Computes aggregate features and initializes tensor statistics

#### **4. Statistical Computation Methods**

**Activity Processing:**
- **`unfolded_extended_activity`**: Handles both convolutional and linear next modules
  - Conv path: Uses `torch.nn.functional.unfold` with bias extension
  - Linear path: Direct activity processing with bias concatenation

**Tensor Statistics:**
- **`compute_s_update()`**: Second-order statistics computation
- **`compute_previous_s_update()`**: Previous layer statistics
- **`compute_previous_m_update()`**: Gradient-activity cross-correlation
- **`construct_full_activity()`**: Multi-module activity aggregation

#### **5. Dynamic Size Management**
**`update_size()`**: Handles tensor reallocation when module configurations change
- ✅ Updates channel dimensions from previous modules
- ✅ Reallocates tensor statistics when shapes change
- ✅ Handles cleanup when no previous modules exist

---

## 🧪 **Comprehensive Test Suite**

### **Test Coverage Achievements**
- **19 test methods** specifically for `Conv2dMergeGrowingModule`
- **100% diff coverage** for all newly introduced code
- **Comprehensive edge case coverage** including error conditions

### **Test Categories**

#### **1. Constructor & Basic Properties**
- Parameter validation and type conversion
- Property getter functionality
- Integration with parent class

#### **2. Module Connection Tests**
- Valid and invalid module type combinations
- Kernel size compatibility validation
- Channel dimension matching
- Warning generation for edge cases

#### **3. Activity Processing Tests**
- Convolutional vs linear next module branches
- Bias handling in both configurations
- Tensor shape validation across different paths

#### **4. Statistical Computation Tests**
- Full activity construction with multiple previous modules
- Tensor statistic updates (S, M, previous S/M)
- Gradient computation integration
- Shape invariant validation

#### **5. Error Handling & Edge Cases**
- Type assertion errors
- Channel mismatch handling
- Missing module warnings
- Tensor reallocation scenarios

---

## 🔨 **Supporting Infrastructure Changes**

### **Enhanced Base Classes**
**`src/gromo/modules/growing_module.py`** (152 insertions, 54 deletions):
- Enhanced `MergeGrowingModule` base class functionality
- Improved tensor statistic management
- Better integration points for merge modules

### **Utility Functions**
**`src/gromo/utils/tools.py`** (81 insertions):
- New utility functions for merge module operations
- Mathematical helpers for tensor computations
- Enhanced statistical computation tools

### **Code Refactoring**
**`src/gromo/modules/linear_growing_module.py`** (89 deletions):
- Code moved to base classes for better reusability
- Elimination of duplicate functionality

### **Enhanced Conv2dGrowingModule**
**Additional properties added:**
- `stride`: Stride property access
- `out_width`, `out_height`: Spatial dimension calculation
- `input_volume`: Input volume computation
- `out_features`, `in_features`: Feature dimension properties

---

## 📚 **Documentation & Testing**

### **Documentation Updates**
- **`docs/source/whats_new.rst`**: Release notes updated
- **`merge-conv-test-plan.md`**: Comprehensive testing strategy (156 lines)

### **Test Infrastructure**
- **`tests/test_conv2d_growing_module.py`**: 595+ lines of merge module tests
- **`tests/test_growing_module.py`**: 160+ lines of base class tests
- **`tests/test_tools.py`**: 215+ lines of utility function tests
- **`tests/test_utils.py`**: Additional utility tests

---

## 🎯 **Technical Highlights**

### **1. Sophisticated Activity Processing**
The module handles two distinct computation paths:
- **Convolutional path**: Uses tensor unfolding for spatial convolutions
- **Linear path**: Direct tensor processing for fully connected operations

### **2. Dynamic Module Validation**
Comprehensive validation system ensuring:
- Type compatibility across connected modules
- Geometric consistency (kernel sizes, channel dimensions)
- Proper statistical tensor initialization

### **3. Flexible Architecture**
- Supports both `Conv2dGrowingModule` and `LinearGrowingModule` as next modules
- Handles multiple previous modules with aggregated statistics
- Automatic property derivation from connected modules

### **4. Robust Error Handling**
- Graceful degradation with warning messages
- Comprehensive assertion-based validation
- Clear error messages for debugging

---

## 🔍 **Code Quality Metrics**

### **Coverage Achievement**
- **100% diff coverage** for all newly introduced lines
- **Zero missing lines** in the core implementation
- **Comprehensive branch coverage** including error paths

### **Test Quality**
- **Deterministic tests** with proper seeding
- **Device-aware testing** using `global_device()`
- **Isolated test cases** without shared state mutation
- **Shape and type validation** over brittle value comparisons

### **Code Organization**
- **Clean separation** of concerns across methods
- **Consistent naming** and documentation
- **Type hints** throughout the implementation
- **Comprehensive docstrings** with parameter descriptions

---

## 🚀 **Integration Benefits**

### **1. Network Growth Capability**
Enables dynamic neural network architectures where multiple convolutional paths can be merged into unified computational units.

### **2. Statistical Tracking**
Maintains comprehensive statistics for both forward and backward passes, enabling advanced optimization techniques.

### **3. Flexible Module Composition**
Supports complex network topologies with multiple input and output connections.

### **4. Performance Optimization**
Efficient tensor operations with proper memory management and shape optimization.

---

## 📋 **Commit History Summary**

The PR evolved through multiple phases:
1. **Initial implementation** of core functionality
2. **Comprehensive testing** with coverage improvements
3. **Integration enhancements** with `LinearGrowingModule` support
4. **Coverage optimization** achieving 100% diff coverage
5. **Final refinements** and documentation updates

Key commits:
- `feat: enhance Conv2dMergeGrowingModule with LinearGrowingModule support`
- `test: add targeted tests to improve diff coverage from 92.18% to 93.16%+`
- `test: add test for input_volume property to achieve 100% diff coverage`

---

## ✅ **Quality Assurance**

### **All Tests Passing**
- 89 total tests across the module
- 19 specific tests for `Conv2dMergeGrowingModule`
- Zero test failures or errors

### **Code Standards**
- **Black formatting** applied
- **isort** import organization
- **Pre-commit hooks** passing
- **Type checking** with proper annotations

### **Documentation**
- Comprehensive docstrings
- Clear parameter descriptions
- Usage examples in tests
- Integration documentation

---

## 🎉 **Conclusion**

This PR successfully implements the `Conv2dMergeGrowingModule` class, providing a robust foundation for dynamic neural network growth in the GROMO framework. The implementation features:

- **Complete functionality** with 100% test coverage
- **Robust architecture** supporting multiple module types
- **Comprehensive validation** and error handling
- **High code quality** with proper documentation and testing
- **Performance optimization** with efficient tensor operations

The implementation is ready for production use and provides a solid foundation for advanced neural network architectures with dynamic growth capabilities.

File: src/gromo/modules/conv2d_growing_module.py [hunk 7]
```diff
@@ -118,6 +531,39 @@ def padding(self):
     def padding(self, value):
         self.layer.padding = value
 
+    @property
+    def stride(self) -> tuple[int, int]:
+        return self.layer.stride
+
+    def __out_dimention(self, dim: int) -> int:
```
[the remark this answers is not in the document]
Comment (stephane-rivaud) [reply src/gromo/modules/conv2d_growing_module.py:538]: Refactoring code to fix this typo in the next commit.

File: src/gromo/modules/conv2d_growing_module.py [hunk 8]
```diff
@@ -14,11 +14,425 @@
...
+        warn(f"Using property in_features in {self}")
+        return self.in_channels
+
+    @property
+    def out_features(self) -> int:
+        return self.input_volume
+
+    @property
+    def output_size(self) -> tuple[int, int]:
+        return self.input_size  # TODO: check for exceptions!
```
[the remark this answers is not in the document]
Comment (stephane-rivaud) [reply src/gromo/modules/conv2d_growing_module.py:81]: Keeping code as is, as I do not know what was the likely exceptions that triggered this TODO comment.

File: src/gromo/modules/conv2d_growing_module.py [hunk 9]
```diff
@@ -14,11 +14,425 @@
     compute_output_shape_conv,
     create_bordering_effect_convolution,
 )
-from gromo.utils.utils import global_device
 
 
 class Conv2dMergeGrowingModule(MergeGrowingModule):
```
[the remark this answers is not in the document]
Comment (stephane-rivaud) [reply src/gromo/modules/conv2d_growing_module.py:19]: I did not understand

File: src/gromo/modules/conv2d_growing_module.py [hunk 10]
```diff
@@ -14,11 +14,428 @@
...
+            next_kernel_size = (next_kernel_size, next_kernel_size)
+        self.kernel_size = next_kernel_size
+        super(Conv2dMergeGrowingModule, self).__init__(
+            post_merge_function=post_merge_function,
+            previous_modules=previous_modules,
+            next_modules=next_modules,
+            allow_growing=allow_growing,
+            tensor_s_shape=(
+                in_channels * next_kernel_size[0] * next_kernel_size[1] + self.use_bias,
+                in_channels * next_kernel_size[0] * next_kernel_size[1] + self.use_bias,
```
[the remark this answers is not in the document]
Comment (stephane-rivaud) [reply src/gromo/modules/conv2d_growing_module.py:49]: Applying copilot suggestion

File: src/gromo/modules/conv2d_growing_module.py [hunk 11]
```diff
@@ -14,11 +14,428 @@
...
+
+    @property
+    def dilation(self) -> int:
+        if len(self.next_modules) <= 0:
+            warn(
+                f"Cannot derive the dilation of Conv2dMergeGrowingModule without setting at least one next module"
+            )
+            return 1
+        elif isinstance(self.next_modules[0], Conv2dGrowingModule):
+            return self.next_modules[0].layer.dilation
```
[the remark this answers is not in the document]
Comment (stephane-rivaud) [reply src/gromo/modules/conv2d_growing_module.py:119]: This issue will be fixed by imposing tuple of (int, int) output instead of int.

File: tests/test_conv2d_growing_module.py [hunk 12]
```diff
@@ -5,16 +5,597 @@
...
+        with self.assertRaises(AssertionError):
+            m.compute_s_update()
+
+    def test_compute_s_update_not_implemented_next_module(self):
+        """Test NotImplementedError for unsupported next module types in compute_s_update."""
+        m = self.merge
+        # Set up valid activity - use 2D to avoid tensor concatenation issues in unfolded_extended_activity
+        m.store_activity = True
+        m.activity = torch.randn(
+            self.batch, self.merge.out_features, device=global_device()
```
[the remark this answers is not in the document]
Comment (stephane-rivaud) [reply tests/test_conv2d_growing_module.py:359]: Applying suggestion from Copilot

File: src/gromo/modules/conv2d_growing_module.py [hunk 13]
```diff
@@ -14,11 +14,382 @@
...
+            )
+            return -1
+        return self.previous_modules[0].out_features
+
+    @property
+    def out_channels(self) -> int:
+        return self.in_channels
+
+    @property
+    def in_features(self) -> int:
```
[the remark this answers is not in the document]
Comment (stephane-rivaud) [reply src/gromo/modules/conv2d_growing_module.py:73]: This renaming would make sense from the convolutional layers point of view. However, `in_features` and `out_features` are methods implemented in `GrowingModule` so the naming makes behavior consistent.

I suggest to keep code as is and discuss proper nomenclature in a code meeting.

File: src/gromo/modules/conv2d_growing_module.py [hunk 14]
```diff
@@ -14,11 +14,430 @@
...
+        warn(f"Using property in_features in {self}")
+        return self.in_channels
+
+    @property
+    def out_features(self) -> int:
+        return self.input_volume
```
[the remark this answers is not in the document]
Comment (stephane-rivaud) [reply src/gromo/modules/conv2d_growing_module.py:79]: I will dig into this to find out if we can have a minimal number of additional method.

File: src/gromo/modules/conv2d_growing_module.py [hunk 15]
```diff
@@ -118,6 +536,39 @@ def padding(self):
...
+    @property
+    def out_width(self) -> int:
+        return self.__out_dimension(0)
+
+    @property
+    def out_height(self) -> int:
+        return self.__out_dimension(1)
+
+    @property
+    def input_volume(self) -> int:
```
[the remark this answers is not in the document]
Comment (stephane-rivaud) [reply src/gromo/modules/conv2d_growing_module.py:561]: Need to dig into it

File: src/gromo/modules/conv2d_growing_module.py [hunk 16]
```diff
@@ -118,6 +536,39 @@ def padding(self):
...
+        return self.in_channels * self.input_size[0] * self.input_size[1]
+
+    @property
+    def out_features(self) -> int:
+        return self.out_channels * self.out_width * self.out_height
+
+    @property
+    def in_features(self) -> int:
```
[the remark this answers is not in the document]
Comment (stephane-rivaud) [reply src/gromo/modules/conv2d_growing_module.py:569]: I need to dig into this as commented above

File: src/gromo/modules/conv2d_growing_module.py [hunk 17]
```diff
@@ -14,11 +14,430 @@
...
+        self.total_out_features = 0
+        for module in self.previous_modules:
+            # Type compatibility already checked above - all are Conv2dGrowingModule
+            if module.out_channels != self.in_channels:
+                raise ValueError(
+                    "The input channels must match the output channels of the previous modules."
+                )
+            self.total_in_features += module.in_features + module.use_bias
+            self.total_out_features += module.out_features
+            # self.total_out_features += module.use_bias
```
[the remark this answers is not in the document]
Comment (stephane-rivaud) [reply src/gromo/modules/conv2d_growing_module.py:266]: Removing inline comment

File: src/gromo/modules/conv2d_growing_module.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (stephane-rivaud) [reply src/gromo/modules/conv2d_growing_module.py:561]: I think that using `input_volume` is fine since it is a term referring to convolution input shapes. Renaming it to `input_features` might be confusing with `in_features` which is equal to `fan_in`. However, we should not rename `in_features` to `fan_in` in this PR to avoid modifying `LinearGrowingModule` class in this PR. Proper naming conventions should be discussed further with the team.

File: src/gromo/modules/conv2d_growing_module.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (stephane-rivaud) [reply src/gromo/modules/conv2d_growing_module.py:569]: This would be a good idea, however, we should also replace this name in `LinearGrowingModule` and should be done in a separate PR, probably after an in-depth discussion about naming convention with the team.

File: src/gromo/modules/conv2d_growing_module.py [hunk 18]
```diff
@@ -14,11 +14,430 @@
...
+        warn(f"Using property in_features in {self}")
+        return self.in_channels
+
+    @property
+    def out_features(self) -> int:
+        return self.input_volume
+
+    @property
+    def output_size(self) -> tuple[int, int]:
+        return self.input_size  # TODO: check for exceptions!
```
[the remark this answers is not in the document]
Comment (stelladk) [reply src/gromo/modules/conv2d_growing_module.py:76]: This comment is here because I am not sure if there is a case where the Conv2dMergeGrowingModule can change the output dimensions because of a post_merge_function for example.

File: src/gromo/modules/growing_module.py [hunk 19]
```diff
@@ -423,9 +444,19 @@ def sum_in_features(self, with_bias: bool = False) -> int:
...
+                ]
             )
-        return np.sum([module.in_features for module in self.previous_modules])
+        return np.sum(
+            [
+                module.in_features
+                for module in self.previous_modules
+                if not isinstance(module, MergeGrowingModule)
+            ]
+        )
```
Comment (TheoRudkiewicz) [inline src/gromo/modules/growing_module.py:459]: I think we do not need `numpy` for this:
```suggestion
            return sum(
                       module.in_features + module.use_bias
                       for module in self.previous_modules
                       if not isinstance(module, MergeGrowingModule)
            )
        return sum(
                module.in_features
                for module in self.previous_modules
                if not isinstance(module, MergeGrowingModule)
        )
```

File: src/gromo/modules/conv2d_growing_module.py [hunk 20]
```diff
@@ -120,6 +538,85 @@ def padding(self):
...
+
+        Returns
+        -------
+        int
+            total number of input elements
+        """
+        return self.in_channels * self.input_size[0] * self.input_size[1]
+
+    @property
+    def out_features(self) -> int:
```
Comment (TheoRudkiewicz) [inline src/gromo/modules/conv2d_growing_module.py:588]: I would advocate to name it `out_volume`, following the same logic as `input_volume`.

File: src/gromo/modules/conv2d_growing_module.py [hunk 21]
```diff
@@ -120,6 +538,85 @@ def padding(self):
...
+        return self.layer.stride
+
+    def __out_dimension(self, dim: int) -> int:
+        return (
+            int(
+                (self.input_size[dim] - self.kernel_size[dim] + 2 * self.padding[dim])
+                / self.stride[dim]
+            )
+            + 1
+        )
```
Comment (TheoRudkiewicz) [inline src/gromo/modules/conv2d_growing_module.py:552]: We could use `gromo.utils.tools.compute_output_shape_conv` to do the computation. (I you commit the suggestion it wont work as I did not put the import.)

```suggestion
        return compute_output_shape_conv(self.input_size, self.layer)
```

Comment (stelladk) [reply src/gromo/modules/conv2d_growing_module.py:552]: I would prefer that there were an option on this function to skip all the checks

**Comment by TheoRudkiewicz** [comment]
I created a separate PR (#133 ) to isolate the modification on `GrowingContainer`.

File: src/gromo/modules/conv2d_growing_module.py (hunk 21 above)
Comment (stelladk) [reply src/gromo/modules/conv2d_growing_module.py:560]: I suggest we do this in a separate PR

**Comment by stelladk** [comment]
I suggest we remove the markdown file and merge this PR

**Comment by sylvchev** [comment]
> I suggest we remove the markdown file and merge this PR

Good call

**Comment by sylvchev** [comment]
Kudos @stelladk for your patient works along those months. Thanks @stephane-rivaud and @TheoRudkiewicz for your help on this challengingly long PR!
